File: js/test/entity/pull/PullEntity.test.js

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

const Path = require('node:path')
const Fs = require('node:fs')

const { test, describe, afterEach } = require('node:test')
const assert = require('node:assert')


const { GithubSDK, BaseFeature, stdutil, config } = require('../../..')

const {
  envOverride,
  liveClientOptions,
  liveDelay,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
} = require('../../utility')


describe('PullEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when GITHUB_TEST_LIVE=TRUE.
  afterEach(liveDelay('GITHUB_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = GithubSDK.test()
    const ent = testsdk.Pull()
    assert(null != ent)
  })


  test('basic', async () => {

    const setup = basicSetup()
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select


    // CREATE
    const pull_ref01_ent = client.Pull()
    let pull_ref01_data = setup.data.new.pull['pull_ref01']
    pull_ref01_data['commit_sha'] = setup.idmap['commit_sha01']
    pull_ref01_data['owner'] = setup.idmap['owner01']
    pull_ref01_data['repo'] = setup.idmap['repo01']

    pull_ref01_data = (await pull_ref01_ent.create(pull_ref01_data)).data()
    assert(null != pull_ref01_data.id)


    // LIST
    const pull_ref01_match = {}
    pull_ref01_match['commit_sha'] = setup.idmap['commit_sha01']
    pull_ref01_match['owner'] = setup.idmap['owner01']
    pull_ref01_match['repo'] = setup.idmap['repo01']

    const pull_ref01_list = (await pull_ref01_ent.list(pull_ref01_match)).map((e) => e.data())

    assert(!isempty(select(pull_ref01_list, { id: pull_ref01_data.id })))


    // UPDATE
    const pull_ref01_data_up0 = {}
    pull_ref01_data_up0.id = pull_ref01_data.id
    pull_ref01_data_up0 ['owner'] = setup.idmap['owner']
    pull_ref01_data_up0 ['repo'] = setup.idmap['repo']

    const pull_ref01_markdef_up0 = { name: 'active_lock_reason', value: 'Mark01-pull_ref01_' + setup.now }
    pull_ref01_data_up0 [pull_ref01_markdef_up0.name] = pull_ref01_markdef_up0.value

    const pull_ref01_resdata_up0 = (await pull_ref01_ent.update(pull_ref01_data_up0)).data()
    assert(pull_ref01_resdata_up0.id === pull_ref01_data_up0.id)

    assert(pull_ref01_resdata_up0[pull_ref01_markdef_up0.name] === pull_ref01_markdef_up0.value)


    // LOAD
    const pull_ref01_match_dt0 = {}
    pull_ref01_match_dt0.id = pull_ref01_data.id
    const pull_ref01_data_dt0 = (await pull_ref01_ent.load(pull_ref01_match_dt0)).data()
    assert(pull_ref01_data_dt0.id === pull_ref01_data.id)


    // REMOVE
    const pull_ref01_match_rm0 = {}
    pull_ref01_match_rm0.id = pull_ref01_data.id
    await pull_ref01_ent.remove(pull_ref01_match_rm0)
  

    // LIST
    const pull_ref01_match_rt0 = {}
    pull_ref01_match_rt0['commit_sha'] = setup.idmap['commit_sha01']
    pull_ref01_match_rt0['owner'] = setup.idmap['owner01']
    pull_ref01_match_rt0['repo'] = setup.idmap['repo01']

    const pull_ref01_list_rt0 = (await pull_ref01_ent.list(pull_ref01_match_rt0)).map((e) => e.data())

    assert(isempty(select(pull_ref01_list_rt0, { id: pull_ref01_data.id })))


  })
})



function basicSetup(extra) {
  // TODO: fix test def options
  const options = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname,
      '../../../../.sdk/test/entity/pull/PullTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = GithubSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['pull01','pull02','pull03','repo01','repo02','repo03','repo01','repo02','repo03','commit01','commit02','commit03','repo01','repo02','repo03','comment01','comment02','comment03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'GITHUB_TEST_PULL_ENTID': idmap,
    'GITHUB_TEST_LIVE': 'FALSE',
    'GITHUB_TEST_EXPLAIN': 'FALSE',
    'GITHUB_APIKEY': '',
  })

  idmap = env['GITHUB_TEST_PULL_ENTID']

  if ('TRUE' === env.GITHUB_TEST_LIVE) {
    client = new GithubSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
        apikey: env.GITHUB_APIKEY,
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when
      // the last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey and
      // server values above and handed the SDK undefined.
      extra || {}
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.GITHUB_TEST_EXPLAIN,
    now: Date.now(),
  }

  return setup
}
  

File: go/test/pull_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/github-sdk/go"
	"github.com/voxgig-sdk/github-sdk/go/core"

	vs "github.com/voxgig-sdk/github-sdk/go/utility/struct"
)

func TestPullEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Pull(nil)
		if ent == nil {
			t.Fatal("expected non-nil PullEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"pull": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Pull(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Pull(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := pullBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "pull." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_PULL_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		pullRef01Ent := client.Pull(nil)
		pullRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "pull"}), "pull_ref01"))
		pullRef01Data["owner"] = setup.idmap["owner01"]
		pullRef01Data["repo"] = setup.idmap["repo01"]

		pullRef01DataResult, err := pullRef01Ent.Create(pullRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		pullRef01Data = core.ToMapAny(entityData(pullRef01DataResult))
		if pullRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if pullRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		pullRef01Match := map[string]any{
			"owner": setup.idmap["owner01"],
			"repo": setup.idmap["repo01"],
		}

		pullRef01ListResult, err := pullRef01Ent.List(pullRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		pullRef01List, pullRef01ListOk := pullRef01ListResult.([]any)
		if !pullRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", pullRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(pullRef01List), map[string]any{"id": pullRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		pullRef01DataUp0Up := map[string]any{
			"id": pullRef01Data["id"],
			"owner": setup.idmap["owner"],
			"repo": setup.idmap["repo"],
		}

		pullRef01MarkdefUp0Name := "active_lock_reason"
		pullRef01MarkdefUp0Value := fmt.Sprintf("Mark01-pull_ref01_%d", setup.now)
		pullRef01DataUp0Up[pullRef01MarkdefUp0Name] = pullRef01MarkdefUp0Value

		pullRef01ResdataUp0Result, err := pullRef01Ent.Update(pullRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		pullRef01ResdataUp0 := core.ToMapAny(entityData(pullRef01ResdataUp0Result))
		if pullRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if pullRef01ResdataUp0["id"] != pullRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if pullRef01ResdataUp0[pullRef01MarkdefUp0Name] != pullRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", pullRef01MarkdefUp0Name, pullRef01ResdataUp0[pullRef01MarkdefUp0Name])
		}

		// LOAD
		pullRef01MatchDt0 := map[string]any{
			"id": pullRef01Data["id"],
		}
		pullRef01DataDt0Loaded, err := pullRef01Ent.Load(pullRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		pullRef01DataDt0LoadResult := core.ToMapAny(entityData(pullRef01DataDt0Loaded))
		if pullRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if pullRef01DataDt0LoadResult["id"] != pullRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func pullBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "pull", "PullTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read pull test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse pull test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"pull01", "pull02", "pull03", "repo01", "repo02", "repo03", "owner01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITHUB_TEST_PULL_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITHUB_TEST_PULL_ENTID": idmap,
		"GITHUB_TEST_LIVE":      "FALSE",
		"GITHUB_TEST_EXPLAIN":   "FALSE",
		"GITHUB_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["GITHUB_TEST_PULL_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add owner alias for update test.
	if idmapResolved["owner"] == nil {
		idmapResolved["owner"] = idmapResolved["owner01"]
	}
	// Add repo alias for update test.
	if idmapResolved["repo"] == nil {
		idmapResolved["repo"] = idmapResolved["repo01"]
	}

	if env["GITHUB_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["GITHUB_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewGithubSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITHUB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITHUB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
